### PR TITLE
1175 Make it possible to create multiple words with the same text

### PR DESCRIPTION
**Pull request by tomaszsmy**

### CODE REVIEW

**Comment by nya-elimuai** [comment]
Thank you, @tomaszsmy. Your code looks great to me 🙂

Can you also add some unit tests for this use-case in `WordDaoTest.java`?

File: src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java [hunk 1]
```diff
@@ -24,6 +25,20 @@ public Word readByText(String text) throws DataAccessException {
...
+        String select = "SELECT w FROM Word w WHERE w.text = :text AND w.wordType ";
+        try {
+            return (Word) em.createQuery(
+                            select + (wordType == null ? "IS NULL AND :wordType IS NULL" : "= :wordType"))
+                    .setParameter("text", text)
+                    .setParameter("wordType", wordType)
+                    .getSingleResult();
+        } catch (NoResultException e) {
+            return null;
+        }
```
Comment (nya-elimuai) [inline src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java:39]:
```suggestion
        String query = "SELECT w " +
                        "FROM Word w " +
                        "WHERE w.text = :text ";
        if (wordType == null) {
            query += "AND w.wordType IS NULL AND :wordType IS NULL";
        } else {
            query += "AND w.wordType = :wordType";
        }
        try {
            return (Word) em.createQuery(query)
                    .setParameter("text", text)
                    .setParameter("wordType", wordType)
                    .getSingleResult();
        } catch (NoResultException e) {
            return null;
        }
```

File: src/test/java/ai/elimu/dao/WordDaoTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration("file:src/main/webapp/WEB-INF/spring/applicationContext-jpa.xml")
```
Comment (nya-elimuai) [inline src/test/java/ai/elimu/dao/WordDaoTest.java:16]: Ref https://github.com/elimu-ai/webapp/pull/1601/files#diff-be21626776e7ac041ba6fad0d43302d787d5245997b8b325d8c983f837ac19a4

```suggestion
@ContextConfiguration(locations={
    "file:src/main/webapp/WEB-INF/spring/applicationContext.xml",
    "file:src/main/webapp/WEB-INF/spring/applicationContext-jpa.xml"
})
```

**Review by nya-elimuai — DISMISSED** [review]

File: src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java [hunk 3]
```diff
@@ -24,6 +25,26 @@ public Word readByText(String text) throws DataAccessException {
         }
     }
 
+    @Override
+    public Word readByTextAndType(String text, WordType wordType) throws DataAccessException {
+        String query = "SELECT w " +
+                "FROM Word w " +
+                "WHERE w.text = :text ";
+        if (wordType == null) {
+            query += "AND w.wordType IS NULL AND :wordType IS NULL";
```
Comment (nya-elimuai) [inline src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java:34]:
```suggestion
            query += "AND w.wordType IS NULL";
```

**Review by nya-elimuai — CHANGES_REQUESTED** [review]
Suggested removal of code in the JPQL query.

File: src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java (hunk 3 above)
Comment (tomaszsmy) [reply src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java:34]: .setParameter("wordType", wordType) requires this parameter in JPQL query.

Comment (nya-elimuai) [reply src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java:34]: > .setParameter("wordType", wordType) requires this parameter in JPQL query.
> 

I see. Can we solve it like this?:
```java
if (wordType != null) {
    query.setParameter("wordType", wordType);
}
```

File: src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java [hunk 4]
```diff
@@ -27,19 +28,24 @@ public Word readByText(String text) throws DataAccessException {
 
     @Override
     public Word readByTextAndType(String text, WordType wordType) throws DataAccessException {
+        Query queryBuilder;
         String query = "SELECT w " +
                 "FROM Word w " +
                 "WHERE w.text = :text ";
+
```
Comment (nya-elimuai) [inline src/main/java/ai/elimu/dao/jpa/WordDaoJpa.java:35]:
```suggestion
```

**Review by nya-elimuai — DISMISSED** [review]

**Comment by nya-elimuai** [comment]
@tomaszsmy Ready to be merged 🙂

**Comment by tomaszsmy** [comment]
> @tomaszsmy Ready to be merged 🙂

remote: error: At least 2 approving reviews are required by reviewers with write access.

File: src/test/java/ai/elimu/dao/WordDaoTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    private Word getWordWitTheSameText(WordType wordType) {
+        Word word = new Word();
+        word.setText(WORD_TEXT);
+        word.setWordType(wordType);
+        return word;
+    }
+
+}
```
Comment (jo-elimu) [inline src/test/java/ai/elimu/dao/WordDaoTest.java:55]:
```suggestion
}

```

**Review by jo-elimu — APPROVED** [review]

**Review by nya-elimuai — APPROVED** [review]

**Pull request merged**